### PR TITLE
chore: data plane changes for credentials in transformation

### PR DESCRIPTION
**Pull request by kanishkkatara**
# Description
Pass credentials along with events to Rudder Transformer, retrieved from the workspace configuration.

https://www.notion.so/rudderstacks/Transformation-secrets-5b234a813aa44e71948a68dffb0a66d2?pvs=4#89cddfd45a5343239e8d5398c7e61463

## Linear Ticket
Fixes DAT-1204
https://linear.app/rudderstack/issue/DAT-1204/data-plane-flow

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
  - Introduced support for managing credentials within various configurations and processes.
  - Added the ability to include credentials in transformer events.

- **Tests**
  - Updated test cases to validate the new credentials functionality.

- **Chores**
  - Enhanced sample backend configuration data to include credentials for testing purposes.

### CODE REVIEW

File: processor/processor.go [hunk 1]
```diff
@@ -836,6 +840,18 @@ func (proc *Handle) backendConfigSubscriber(ctx context.Context) {
...
+	var credentialsList []backendconfig.Credential
+	for _, credential := range credentialsMap {
+		credentialsList = append(credentialsList, backendconfig.Credential{
+			Key:      credential.Key,
+			Value:    credential.Value,
+			IsSecret: credential.IsSecret,
+		})
+	}
+	return credentialsList
+}
```
Comment (abhimanyubabbar) [inline processor/processor.go:853]: We are loosing the credential identifier in this process and only sending list of k/v combinations. Why are we doing that ?

We should contain necessary information like identifier for debugging purposes as well. Wdyt ?!

**Review by abhimanyubabbar — COMMENTED** [review]
Can you add a testcase to make sure the backend credentials will make it to the request to transformer in rudder-server ?!

**Comment by abhimanyubabbar** [comment]
Can we add tests for the change ?

File: processor/processor.go [hunk 2]
```diff
@@ -144,6 +144,7 @@ type Handle struct {
 		eventSchemaV2Enabled            bool
 		archivalEnabled                 config.ValueLoader[bool]
 		eventAuditEnabled               map[string]bool
+		credentials                     map[string][]transformer.Credential
```
Comment (achettyiitr) [inline processor/processor.go:147]:
```suggestion
		credentialsMap                     map[string][]transformer.Credential
```

File: processor/processor.go [hunk 3]
```diff
@@ -836,6 +840,19 @@ func (proc *Handle) backendConfigSubscriber(ctx context.Context) {
...
+	var credentialsList []transformer.Credential
+	for id, credential := range credentialsMap {
+		credentialsList = append(credentialsList, transformer.Credential{
+			ID:       id,
+			Key:      credential.Key,
+			Value:    credential.Value,
+			IsSecret: credential.IsSecret,
+		})
+	}
+	return credentialsList
```
Comment (achettyiitr) [inline processor/processor.go:853]:
```suggestion
	return lo.MapToSlice(credentialsMap, func(key string, value backendconfig.Credential) transformer.Credential {
		return transformer.Credential{
			ID:       key,
			Key:      value.Key,
			Value:    value.Value,
			IsSecret: value.IsSecret,
		}
	})
```

File: processor/processor.go [hunk 4]
```diff
@@ -819,6 +821,7 @@ func (proc *Handle) backendConfigSubscriber(ctx context.Context) {
 			}
 			workspaceLibrariesMap[workspaceID] = wConfig.Libraries
 			eventAuditEnabled[workspaceID] = wConfig.Settings.EventAuditEnabled
+			credentialsMap[workspaceID] = ConvertMapToList(wConfig.Credentials)
```
Comment (achettyiitr) [inline processor/processor.go:824]:
```suggestion
			credentialsMap[workspaceID] = lo.MapToSlice(wConfig.Credentials, func(key string, value backendconfig.Credential) transformer.Credential {
				return transformer.Credential{
					ID:       key,
					Key:      value.Key,
					Value:    value.Value,
					IsSecret: value.IsSecret,
				}
			})
```

File: processor/processor.go [hunk 5]
```diff
@@ -842,6 +853,19 @@ func (proc *Handle) backendConfigSubscriber(ctx context.Context) {
 	}
 }
 
+func ConvertMapToList(credentialsMap map[string]backendconfig.Credential) []transformer.Credential {
```
Comment (achettyiitr) [inline processor/processor.go:856]: We can remove this.

Comment (kanishkkatara) [reply processor/processor.go]: done

**Comment by achettyiitr** [comment]
> Can you add a testcase to make sure the backend credentials will make it to the request to transformer in rudder-server ?!

We can write contract tests for the transformer. Something like this would be helpful? https://github.com/rudderlabs/rudder-server/pull/4787

File: integration_test/transformer_contract/transformer_contract_test.go [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+		require.NoError(t, err)
+
+		requireJobsCount(t, ctx, postgresContainer.DB, "gw", jobsdb.Succeeded.State, eventsCount)
+		requireJobsCount(t, ctx, postgresContainer.DB, "rt", jobsdb.Succeeded.State, eventsCount)
+
+		cancel()
+		require.NoError(t, wg.Wait())
+	})
+	t.Run("Dest Transformer", func(t *testing.T) {})
+	t.Run("Tracking Plan Validation", func(t *testing.T) {})
```
Comment (achettyiitr) [inline integration_test/transformer_contract/transformer_contract_test.go:182]: Can we mark it as TODO.

**Review by achettyiitr — APPROVED** [review]

**Review by cisse21 — APPROVED** [review]

**Review by ktgowtham — APPROVED** [review]

File: backend-config/types.go [hunk 7]
```diff
@@ -74,6 +74,12 @@ type SourceT struct {
 	}
 }
 
+type Credential struct {
+	Key      string `json:"key"`
+	Value    string `json:"value"`
+	IsSecret bool   `json:"isSecret"`
```
[the remark this answers is not in the document]
Comment (kanishkkatara) [reply backend-config/types.go:80]: Yes, we receive IsSecret from the workspace config, and we need to send it to the transformer

**Review by abhimanyubabbar — APPROVED** [review]

**Pull request merged**